backend/test_observer/controllers/test_results/test_results.py

```diff
@@ -0,0 +1,307 @@
+    ArtefactBuild,
+    Environment,
+    TestExecution,
+    TestResult,
+    TestCase,
+    Issue,
+    IssueTestResultAttachment,
+)
+from test_observer.data_access.models_enums import FamilyName
+from test_observer.data_access.setup import get_db
+
+from .models import TestResultSearchResponse
+
+router = APIRouter(tags=["test-results"])
+
+
+@router.get("", response_model=TestResultSearchResponse)
+def search_test_results(
+    family: str | None = Query(
+        None, description="Filter by artefact family (e.g., Charm, Snap)"
+    ),
+    environment: str | None = Query(
+        None,
+        description="Filter by environment name (e.g., Juju:3/stable ubuntu:20.04)",
+    ),
+    test_case: str | None = Query(
+        None, description="Filter by test case name (e.g., test_deploy)"
+    ),
+    template_id: str | None = Query(None, description="Filter by template ID"),
+    issues: str | None = Query(
+        None, description="Filter by Jira or GitHub issue IDs (comma-separated)"
+    ),
+    from_date: datetime | None = Query(
+        None, description="Filter results from this timestamp"
+    ),
+    until: datetime | None = Query(
+        None, description="Filter results until this timestamp"
+    ),
+    limit: int = Query(
+        50, ge=1, le=1000, description="Maximum number of results to return"
+    ),
+    offset: int = Query(
+        0, ge=0, description="Number of results to skip for pagination"
+    ),
+    db: Session = Depends(get_db),
+) -> TestResultSearchResponse:
+    """
+    Search test results across artefacts using flexible filters.
+
+    This endpoint uses a single optimized query with window functions to get both
+    the total count and paginated results in one database round trip.
+    """
+
+    # Build filters
+    filters = []
+
+    if family:
+        try:
+            family_enum = FamilyName(family.lower())
+            filters.append(Artefact.family == family_enum)
+        except ValueError:
+            filters.append(Artefact.family.is_(None))
+
+    if environment:
+        filters.append(Environment.name.ilike(f"%{environment}%"))
+
+    if test_case:
+        filters.append(TestCase.name.ilike(f"%{test_case}%"))
+
+    if template_id:
+        filters.append(TestCase.template_id == template_id)
+
+    if issues:
+        # Handle multiple issue IDs separated by commas
+        issue_list = [issue.strip() for issue in issues.split(",") if issue.strip()]
+        if issue_list:
+            try:
+                issue_ids = [int(issue_id) for issue_id in issue_list]
+                filters.append(
+                    TestResult.id.in_(
+                        select(IssueTestResultAttachment.test_result_id)
+                        .join(IssueTestResultAttachment.issue)
+                        .where(Issue.id.in_(issue_ids))
+                    )
+                )
+            except ValueError:
+                return TestResultSearchResponse(count=0, test_results=[])
```

Comment:
I believe this is triggered if the issue id isn't an int? I think that should be a bad request error.